pynicotine/portmapper.py

```diff
@@ -41,7 +45,127 @@ def set_port(self, port, local_ip_address):
         self.local_ip_address = local_ip_address
 
 
+class NATPMP(BaseImplementation):
+    """ Implementation of the NAT-PMP protocol
+    https://www.rfc-editor.org/rfc/rfc6886 """
+
+    NAME = "NAT-PMP"
+    REQUEST_PORT = 5351
+    REQUEST_TIMEOUT = 0.250
+
+    class PortmapResponse:
+
+        def __init__(self, message):
+
+            self.message = message
+            (
+                _version,
+                _op_code,
+                self.result,
+                _sec_since_epoch,
+                _private_port,
+                _public_port,
+                _lease_duration
+            ) = struct.unpack('!BBHIHHI', message)
+
+        def __bytes__(self):
+            return self.message
+
+    class PortmapRequest:
+
+        RESERVED_VALUE = 0
+        TCP_OP_CODE = 2
+        VERSION = 0
+
+        def __init__(self, public_port, private_port, lease_duration):
+
+            self._public_port = public_port
+            self._private_port = private_port
+            self._lease_duration = lease_duration
+
+        def sendto(self, sock, addr):
+
+            msg = bytes(self)
+            sock.sendto(msg, addr)
+
+            log.add_debug("NAT-PMP: Portmap request: %s", msg)
+
+        def __bytes__(self):
+
+            return struct.pack(
+                '!BBHHHI',
+                self.VERSION,
+                self.TCP_OP_CODE,
+                self.RESERVED_VALUE,
+                self._public_port,
+                self._private_port,
+                self._lease_duration
+            )
+
+    def __init__(self):
+        super().__init__()
+        self._gateway_address = None
+
+    @staticmethod
+    def _get_gateway_address():
+
+        if sys.platform == "linux":
+            output = subprocess.check_output(["ip", "route", "list"])
+            return output.rsplit(b"default via", maxsplit=1)[-1].split()[0]
+
+        if sys.platform == "win32":
+            gateway_pattern = re.compile(b".*?0.0.0.0[ ]+0.0.0.0[ ]+(.*?)[ ]+?.*?\n")
+        else:
+            gateway_pattern = re.compile(b"(?:default|0\\.0\\.0\\.0|::/0)\\s+([\\w\\.:]+)\\s+.*UG")
+
+        output = subprocess.check_output(["netstat", "-rn"], shell=True)
+        return gateway_pattern.search(output).group(1)
+
+    def _request_port_mapping(self, public_port, private_port, lease_duration):
+
+        with socket.socket(family=socket.AF_INET, type=socket.SOCK_DGRAM, proto=socket.IPPROTO_UDP) as sock:
+            sock.bind((self.local_ip_address, 0))
+            sock.settimeout(self.REQUEST_TIMEOUT)
+
+            request = self.PortmapRequest(public_port, private_port, lease_duration)
+            request.sendto(sock, (self._gateway_address, self.REQUEST_PORT))
+
+            try:
+                response = self.PortmapResponse(message=sock.recv(16))
+                return response.result
+
+            except socket.timeout:
+                pass
+
```

Comment:
What about repeating the request and increasing the timeout, as per the NAT-PMP spec?

Reply:
Since the spec requires the timeout is doubled up to 9 times, the resulting 64 seconds is a long time for the thread to be potentially waiting, as this might might prevent shutdown from occurring as expected in the event of `want_abort` becoming true in the main networking thread, so this would have to be accounted for.

It might be necessary to use the `event.schedule()` method for this purpose, or some other similar solution.

Reply:
@slook
Why not simply use `PortMapper.add_port_mapping(blocking=True)` so that it never blocks? I'm not very familiar with concurrent programming, so please have mercy.

Reply:
It could well be simpler to avoid spawning any new threads or scheduled events that have the potential to go into a zombie state or otherwise cause spurious random behaviour.

My instinct tells me that implementing a retry mechanism as defined in the spec isn't really going be that useful in our use case, because from my experience if a local connection does not succeed the first time then it will never work no matter how many successive attempts are made.

However, for me this is entirely theoretical because I don't know about NAT-PMP and cannot test this module.

Please let us know what happens when you run this branch, and push a PR with any modifications/corrections that are required for it to work properly (i.e. without unrecoverable errors/failures), whilst considering that new functionality can always be added at a later date, since it is better for the implementation to be incomplete rather than incorrect.

Reply:
The reason we block when forwarding ports, is because we want to have an open port before we connect to the server and other peers.

I'm not sure how much of a difference it would make if we follow the spec to the teeth, or just increase the timeout of the first try to, say, 1 second, like we do with UPnP. With my router, the first try has never failed so far.

Reply:
It seems reasonable enough.

Reply:
I implemented a simple retry mechanism and set max attempts to 2, which in total gives 0.75 seconds of timeout. I then ran this branch with my modifications and ensured that everything works as expected, both when port mapping fails and when it succeeds. I also added debug log messages.

@slook You wrote:
> Please let us know what happens when you run this branch, and **push a PR** with any modifications/corrections that are required for it to work properly

Do you want me to push a new PR and close my original one? Or perhaps I can somehow contribute to this PR (#2408) to group the commits nicely.

Reply:
I could merge this PR as-is now, so you can make further changes on top of master in a separate PR, if it sounds fine?

Reply:
@homedirectory I suggest you make a new PR to implement the retry mechanism to add yourself to contributors.

Agreed up to 3 retry attempts would be wise to implement, as this may help account for sleepy network devices and such, without creating undue delay whilst establishing a connection.

Reply:
@mathiascode 
> I could merge this PR as-is now, so you can make further changes on top of master in a separate PR, if it sounds fine?

Makes sense, go ahead.